In [48]:
import pandas as pd

In [49]:
patient_df = pd.read_csv('./raw/patients.csv')
patient_df.head()

subject_id  hadm_id  icustay_id gender              ethnicity        age  \
0           3   145834      211552      M                  WHITE  76.526792   
1           4   185777      294638      F                  WHITE  47.845047   
2           6   107064      228232      F                  WHITE  65.942297   
3           9   150750      220597      M  UNKNOWN/NOT SPECIFIED  41.790228   
4          11   194540      229441      F                  WHITE  50.148295   

  insurance            admittime               diagnosis_at_admission  \
0  Medicare  2101-10-20 19:08:00                          HYPOTENSION   
1   Private  2191-03-16 00:28:00  FEVER,DEHYDRATION,FAILURE TO THRIVE   
2  Medicare  2175-05-30 07:15:00            CHRONIC RENAL FAILURE/SDA   
3  Medicaid  2149-11-09 13:06:00                      HEMORRHAGIC CVA   
4   Private  2178-04-16 06:18:00                           BRAIN MASS   

             dischtime  ...              outtime   los_icu  admission_type  \
0  2101-10-31 13:58:00  ...  2101-10-26 20:43:09  6.064560       EMERGENCY   
1  2191-03-23 18:41:00  ...  2191-03-17 16:46:31  1.678472       EMERGENCY   
2  2175-06-15 16:00:00  ...  2175-06-03 13:39:54  3.672917        ELECTIVE   
3  2149-11-14 10:15:00  ...  2149-11-14 20:52:14  5.323056       EMERGENCY   
4  2178-05-11 19:00:00  ...  2178-04-17 20:21:05  1.584410       EMERGENCY   

   first_careunit  mort_icu mort_hosp  hospital_expire_flag  hospstay_seq  \
0            MICU         0         0                     0             1   
1            MICU         0         0                     0             1   
2            SICU         0         0                     0             1   
3            MICU         1         1                     1             1   
4            SICU         0         0                     0             1   

   readmission_30 max_hours  
0               0       145  
1               0        40  
2               0        88  
3               0       127  
4               0        38  

[5 rows x 31 columns]

In [50]:
# Create total_los column = (dischtime - admittime)/(seconds in a day)
patient_df['admittime'] = pd.to_datetime(patient_df['admittime'])
patient_df['dischtime'] = pd.to_datetime(patient_df['dischtime'])

patient_df['total_los'] = ((patient_df['dischtime'] - patient_df['admittime']).dt.total_seconds() / (24 * 60 * 60)).round(4)
patient_df.head()

subject_id  hadm_id  icustay_id gender              ethnicity        age  \
0           3   145834      211552      M                  WHITE  76.526792   
1           4   185777      294638      F                  WHITE  47.845047   
2           6   107064      228232      F                  WHITE  65.942297   
3           9   150750      220597      M  UNKNOWN/NOT SPECIFIED  41.790228   
4          11   194540      229441      F                  WHITE  50.148295   

  insurance           admittime               diagnosis_at_admission  \
0  Medicare 2101-10-20 19:08:00                          HYPOTENSION   
1   Private 2191-03-16 00:28:00  FEVER,DEHYDRATION,FAILURE TO THRIVE   
2  Medicare 2175-05-30 07:15:00            CHRONIC RENAL FAILURE/SDA   
3  Medicaid 2149-11-09 13:06:00                      HEMORRHAGIC CVA   
4   Private 2178-04-16 06:18:00                           BRAIN MASS   

            dischtime  ...   los_icu  admission_type  first_careunit  \
0 2101-10-31 13:58:00  ...  6.064560       EMERGENCY            MICU   
1 2191-03-23 18:41:00  ...  1.678472       EMERGENCY            MICU   
2 2175-06-15 16:00:00  ...  3.672917        ELECTIVE            SICU   
3 2149-11-14 10:15:00  ...  5.323056       EMERGENCY            MICU   
4 2178-05-11 19:00:00  ...  1.584410       EMERGENCY            SICU   

   mort_icu  mort_hosp hospital_expire_flag  hospstay_seq  readmission_30  \
0         0          0                    0             1               0   
1         0          0                    0             1               0   
2         0          0                    0             1               0   
3         1          1                    1             1               0   
4         0          0                    0             1               0   

   max_hours total_los  
0        145   10.7847  
1         40    7.7590  
2         88   16.3646  
3        127    4.8812  
4         38   25.5292  

[5 rows x 32 columns]

In [51]:
lab_df = pd.read_csv('./raw/vitals_labs_mean.csv')
lab_df.head()

subject_id  hadm_id  icustay_id  hours_in  alanine aminotransferase  \
0           3   145834      211552         0                      25.0   
1           3   145834      211552         1                       NaN   
2           3   145834      211552         2                       NaN   
3           3   145834      211552         3                       NaN   
4           3   145834      211552         4                       NaN   

   albumin  albumin ascites  albumin pleural  albumin urine  \
0      1.8              NaN              NaN            NaN   
1      NaN              NaN              NaN            NaN   
2      NaN              NaN              NaN            NaN   
3      NaN              NaN              NaN            NaN   
4      NaN              NaN              NaN            NaN   

   alkaline phosphate  ...  total protein  total protein urine  troponin-i  \
0                73.0  ...            NaN                  NaN         NaN   
1                 NaN  ...            NaN                  NaN         NaN   
2                 NaN  ...            NaN                  NaN         NaN   
3                 NaN  ...            NaN                  NaN         NaN   
4                 NaN  ...            NaN                  NaN         NaN   

   troponin-t  venous pvo2  weight  white blood cell count  \
0         NaN          NaN     NaN               14.842857   
1         NaN          NaN     NaN                     NaN   
2         NaN          NaN     NaN                     NaN   
3         NaN          NaN     NaN                     NaN   
4         NaN          NaN     NaN                     NaN   

   white blood cell count urine    ph  ph urine  
0                           NaN  7.40       5.0  
1                           NaN   NaN       NaN  
2                           NaN  7.26       NaN  
3                           NaN   NaN       NaN  
4                           NaN   NaN       NaN  

[5 rows x 108 columns]

In [52]:
tmp_features = ['']

In [53]:
print(patient_df.shape)
print(lab_df.shape)

(34472, 32)
(2200954, 108)


### Data Cleaning

First check if there're any patient who does not exists

In [54]:
# Find the difference between subject_id
print(len(set(patient_df['subject_id']) - set(lab_df['subject_id'])))

0


Đầu tiên, chúng tôi phát hiện trong dữ liệu có nhiều bệnh nhân có thời gian bắt đầu điều trị ICU lần đầu tiên diễn ra trước thời gian nhập viện. 

Điều này có thể lí giải là nhiều bệnh nhân nhập viện trong tình trạng khẩn cấp (bị bệnh nặng, tai nạn hoặc thai phụ sinh sản) nên được đưa vào ICU trước khi làm thủ tục nhập viện. 

Chúng tôi giải quyết những trường hợp này bằng cách thay thời gian nhập viện bởi thời gian vào ICU lần đầu tiên.

(Trang 24, 25)

In [55]:
tmp_features = ['subject_id', 'hadm_id', 'icustay_id', 'admittime', 'intime']
tmp = patient_df[patient_df['admittime'] > patient_df['intime']][tmp_features][:10]
tmp

subject_id  hadm_id  icustay_id           admittime               intime
23           35   166707      282039 2122-02-10 11:15:00  2122-02-10 09:39:59
230         347   119310      268893 2118-05-23 13:30:00  2118-05-23 10:56:34
251         373   104540      211665 2198-08-27 12:15:00  2198-08-27 09:17:10
274         409   105471      224525 2159-09-17 12:00:00  2159-09-17 10:33:00
321         487   160958      228805 2130-10-31 12:15:00  2130-10-31 11:04:31
514         796   127253      253353 2137-11-11 12:45:00  2137-11-11 10:23:13
524         810   168366      257899 2136-01-18 21:13:00  2136-01-18 01:10:26
615         957   145966      292666 2162-03-22 11:45:00  2162-03-22 10:42:04
712        1108   108891      208969 2198-04-16 16:00:00  2198-04-16 13:06:00
912        1417   105034      236094 2129-06-17 23:30:00  2129-06-17 22:48:53

In [56]:
tmp.to_csv("./latex_table/4.5.csv", index=False)

Chúng tôi còn quan sát từ dữ liệu có một số bệnh nhân có thời gian nhập viện xảy ra sau thời gian xuất viện. Điều này là vô lý, do đó chúng tôi loại bỏ những bệnh nhân này khỏi bộ dữ liệu.

(Trang 25)

In [57]:
# Admittime > Dischtime
tmp_features = ['subject_id', 'hadm_id', 'icustay_id', 'admittime', 'dischtime']
tmp = patient_df[patient_df['admittime'] > patient_df['dischtime']][tmp_features]

In [58]:
len(tmp)

21

In [59]:
# Drop from patient_df
patient_df = patient_df[~patient_df['subject_id'].isin(tmp['subject_id'])]

In [60]:
tmp = tmp[:15]
tmp

subject_id  hadm_id  icustay_id           admittime           dischtime
120           181   102631      246694 2153-10-12 09:49:00 2153-10-12 06:29:00
1520         2420   135098      227111 2184-12-01 19:28:00 2184-12-01 16:50:00
3376         5567   182804      233453 2181-07-09 15:31:00 2181-07-09 12:00:00
4147         6832   164068      208677 2142-05-23 13:20:00 2142-05-23 12:00:00
5186         8604   102784      218796 2110-01-06 15:15:00 2110-01-06 12:00:00
7232        11983   129454      207034 2138-08-14 14:12:00 2138-08-14 12:00:00
7774        12844   181672      285687 2192-05-07 12:24:00 2192-05-07 12:00:00
11215       18567   143379      202436 2133-02-08 15:24:00 2133-02-08 07:17:00
11559       19124   156524      275239 2139-10-12 04:08:00 2139-10-12 01:13:00
12272       20349   131432      278656 2183-12-10 16:49:00 2183-12-10 07:48:00
12824       21272   156385      291938 2162-09-25 07:36:00 2162-09-25 01:23:00
16671       27558   112191      267222 2177-08-16 12:01:00 2177-08-16 12:00:00
17510       28760   168303      246109 2167-01-14 13:29:00 2167-01-14 12:00:00
17550       28822   187804      225516 2162-04-02 12:33:00 2162-04-02 12:00:00
19414       31644   195619      270822 2139-07-27 23:25:00 2139-07-27 18:30:00

In [61]:
tmp.to_csv("./latex_table/4.6.csv", index=False)

Tiếp theo, chúng tôi phát hiện có 373 bệnh nhân có tổng thời gian nằm viện nhỏ hơn tổng thời gian nằm ICU. 

In [62]:
# Check los_icu > total_los + 0.5
tmp_features = ['subject_id', 'hadm_id', 'icustay_id', 'los_icu', 'total_los']
tmp = patient_df[patient_df['los_icu'] > patient_df['total_los'] + 0.5][tmp_features]
tmp

subject_id  hadm_id  icustay_id   los_icu  total_los
279           417   178013      207928  1.341331     0.3722
427           657   165342      243768  3.455949     2.8062
569           886   130937      294616  2.261806     1.2625
585           908   134650      229578  2.276539     1.6250
602           937   148592      228181  5.253021     3.5562
...           ...      ...         ...       ...        ...
33516       95991   173297      207431  7.487604     3.7083
33915       97609   166990      243152  2.443762     0.9812
34194       98797   105447      244147  1.238171     0.6625
34235       98995   123142      223471  1.945012     0.3944
34461       99955   108494      243255  5.671933     5.1174

[314 rows x 5 columns]

In [63]:
tmp = tmp[:17]
tmp.to_csv("./latex_table/4.7.csv", index=False)

Bước cuối cùng trong phần này, chúng tôi tiến hành loại bỏ những bệnh nhân nằm viện quá 90 ngày. Lí do chúng tôi thực hiện điều này vì số lượng bệnh nhân dạng này rất ít (55 bệnh nhân).
Những bệnh nhân này có thể xem là trường hợp ngoại lệ

(trang 26)

In [64]:
# Drop total_los > 90
patient_df = patient_df[patient_df['total_los'] <= 90]

Và tất nhiên, không quên loại bỏ những bệnh nhân không còn tồn tại nữa trong lab_df

In [65]:
# Drop from lab_df
lab_df = lab_df[lab_df['subject_id'].isin(patient_df['subject_id'])]

## Feature Extraction

húng tôi dựa trên các nghiên cứu trước đó cho bài toán dự đoán số ngày nằm
viện, đặc biệt là các nghiên cứu trên bộ dữ liệu MIMIC-III, các tác giả thường đề xuất sử dụng hai nhóm đặc trưng sau đây:
+ Nhóm đặc trưng các chỉ số xét nghiệm của bệnh nhân.
+ Nhóm đặc trưng các thông tin cơ bản của bệnh nhâ

In [66]:
# Convert patient more than 90 ys to 300
patient_df['age'] = patient_df['age'].apply(lambda x: 300 if x > 90 else x)

In [67]:
patient_df.head()

subject_id  hadm_id  icustay_id gender              ethnicity        age  \
0           3   145834      211552      M                  WHITE  76.526792   
1           4   185777      294638      F                  WHITE  47.845047   
2           6   107064      228232      F                  WHITE  65.942297   
3           9   150750      220597      M  UNKNOWN/NOT SPECIFIED  41.790228   
4          11   194540      229441      F                  WHITE  50.148295   

  insurance           admittime               diagnosis_at_admission  \
0  Medicare 2101-10-20 19:08:00                          HYPOTENSION   
1   Private 2191-03-16 00:28:00  FEVER,DEHYDRATION,FAILURE TO THRIVE   
2  Medicare 2175-05-30 07:15:00            CHRONIC RENAL FAILURE/SDA   
3  Medicaid 2149-11-09 13:06:00                      HEMORRHAGIC CVA   
4   Private 2178-04-16 06:18:00                           BRAIN MASS   

            dischtime  ...   los_icu  admission_type  first_careunit  \
0 2101-10-31 13:58:00  ...  6.064560       EMERGENCY            MICU   
1 2191-03-23 18:41:00  ...  1.678472       EMERGENCY            MICU   
2 2175-06-15 16:00:00  ...  3.672917        ELECTIVE            SICU   
3 2149-11-14 10:15:00  ...  5.323056       EMERGENCY            MICU   
4 2178-05-11 19:00:00  ...  1.584410       EMERGENCY            SICU   

   mort_icu  mort_hosp hospital_expire_flag  hospstay_seq  readmission_30  \
0         0          0                    0             1               0   
1         0          0                    0             1               0   
2         0          0                    0             1               0   
3         1          1                    1             1               0   
4         0          0                    0             1               0   

   max_hours total_los  
0        145   10.7847  
1         40    7.7590  
2         88   16.3646  
3        127    4.8812  
4         38   25.5292  

[5 rows x 32 columns]

In [68]:
col_to_keep = ['subject_id', 'hadm_id', 'icustay_id', 'gender', 'ethnicity', 'age', 'insurance', 'admittime', 'dischtime', 'intime', 'outtime', 'los_icu', 'admission_type', 'first_careunit',  'total_los']

In [69]:
patient_df = patient_df[col_to_keep]
patient_df.head()

subject_id  hadm_id  icustay_id gender              ethnicity        age  \
0           3   145834      211552      M                  WHITE  76.526792   
1           4   185777      294638      F                  WHITE  47.845047   
2           6   107064      228232      F                  WHITE  65.942297   
3           9   150750      220597      M  UNKNOWN/NOT SPECIFIED  41.790228   
4          11   194540      229441      F                  WHITE  50.148295   

  insurance           admittime           dischtime               intime  \
0  Medicare 2101-10-20 19:08:00 2101-10-31 13:58:00  2101-10-20 19:10:11   
1   Private 2191-03-16 00:28:00 2191-03-23 18:41:00  2191-03-16 00:29:31   
2  Medicare 2175-05-30 07:15:00 2175-06-15 16:00:00  2175-05-30 21:30:54   
3  Medicaid 2149-11-09 13:06:00 2149-11-14 10:15:00  2149-11-09 13:07:02   
4   Private 2178-04-16 06:18:00 2178-05-11 19:00:00  2178-04-16 06:19:32   

               outtime   los_icu admission_type first_careunit  total_los  
0  2101-10-26 20:43:09  6.064560      EMERGENCY           MICU    10.7847  
1  2191-03-17 16:46:31  1.678472      EMERGENCY           MICU     7.7590  
2  2175-06-03 13:39:54  3.672917       ELECTIVE           SICU    16.3646  
3  2149-11-14 20:52:14  5.323056      EMERGENCY           MICU     4.8812  
4  2178-04-17 20:21:05  1.584410      EMERGENCY           SICU    25.5292

In [70]:
# Dummy variables
patient_df = pd.get_dummies(patient_df, 
                            columns=['gender', 'ethnicity', 'first_careunit', 'insurance', 'admission_type'],
                            drop_first=True,
                            dtype='uint8')

patient_df.shape

(34414, 61)

In [71]:
# Drop admittime, dischtime, intime, outtime
patient_df.drop(columns=['admittime', 'dischtime', 'intime', 'outtime'], inplace=True)


In [72]:
patient_df.shape

(34414, 57)

In [73]:
patient_df.head()

subject_id  hadm_id  icustay_id        age   los_icu  total_los  gender_M  \
0           3   145834      211552  76.526792  6.064560    10.7847         1   
1           4   185777      294638  47.845047  1.678472     7.7590         0   
2           6   107064      228232  65.942297  3.672917    16.3646         0   
3           9   150750      220597  41.790228  5.323056     4.8812         1   
4          11   194540      229441  50.148295  1.584410    25.5292         0   

   ethnicity_AMERICAN INDIAN/ALASKA NATIVE FEDERALLY RECOGNIZED TRIBE  \
0                                                  0                    
1                                                  0                    
2                                                  0                    
3                                                  0                    
4                                                  0                    

   ethnicity_ASIAN  ethnicity_ASIAN - ASIAN INDIAN  ...  first_careunit_CSRU  \
0                0                               0  ...                    0   
1                0                               0  ...                    0   
2                0                               0  ...                    0   
3                0                               0  ...                    0   
4                0                               0  ...                    0   

   first_careunit_MICU  first_careunit_SICU  first_careunit_TSICU  \
0                    1                    0                     0   
1                    1                    0                     0   
2                    0                    1                     0   
3                    1                    0                     0   
4                    0                    1                     0   

   insurance_Medicaid  insurance_Medicare  insurance_Private  \
0                   0                   1                  0   
1                   0                   0                  1   
2                   0                   1                  0   
3                   1                   0                  0   
4                   0                   0                  1   

   insurance_Self Pay  admission_type_EMERGENCY  admission_type_URGENT  
0                   0                         1                      0  
1                   0                         1                      0  
2                   0                         0                      0  
3                   0                         1                      0  
4                   0                         1                      0  

[5 rows x 57 columns]

In [74]:
patient_df.to_csv('./cleaned/patient.csv', index=False)

In [75]:
lab_df

subject_id  hadm_id  icustay_id  hours_in  alanine aminotransferase  \
0                 3   145834      211552         0                      25.0   
1                 3   145834      211552         1                       NaN   
2                 3   145834      211552         2                       NaN   
3                 3   145834      211552         3                       NaN   
4                 3   145834      211552         4                       NaN   
...             ...      ...         ...       ...                       ...   
2200949       99999   113369      246512        22                       NaN   
2200950       99999   113369      246512        23                       NaN   
2200951       99999   113369      246512        24                       NaN   
2200952       99999   113369      246512        25                       NaN   
2200953       99999   113369      246512        26                       NaN   

         albumin  albumin ascites  albumin pleural  albumin urine  \
0            1.8              NaN              NaN            NaN   
1            NaN              NaN              NaN            NaN   
2            NaN              NaN              NaN            NaN   
3            NaN              NaN              NaN            NaN   
4            NaN              NaN              NaN            NaN   
...          ...              ...              ...            ...   
2200949      NaN              NaN              NaN            NaN   
2200950      NaN              NaN              NaN            NaN   
2200951      NaN              NaN              NaN            NaN   
2200952      NaN              NaN              NaN            NaN   
2200953      NaN              NaN              NaN            NaN   

         alkaline phosphate  ...  total protein  total protein urine  \
0                      73.0  ...            NaN                  NaN   
1                       NaN  ...            NaN                  NaN   
2                       NaN  ...            NaN                  NaN   
3                       NaN  ...            NaN                  NaN   
4                       NaN  ...            NaN                  NaN   
...                     ...  ...            ...                  ...   
2200949                 NaN  ...            NaN                  NaN   
2200950                 NaN  ...            NaN                  NaN   
2200951                 NaN  ...            NaN                  NaN   
2200952                 NaN  ...            NaN                  NaN   
2200953                 NaN  ...            NaN                  NaN   

         troponin-i  troponin-t  venous pvo2  weight  white blood cell count  \
0               NaN         NaN          NaN     NaN               14.842857   
1               NaN         NaN          NaN     NaN                     NaN   
2               NaN         NaN          NaN     NaN                     NaN   
3               NaN         NaN          NaN     NaN                     NaN   
4               NaN         NaN          NaN     NaN                     NaN   
...             ...         ...          ...     ...                     ...   
2200949         NaN         NaN          NaN     NaN                     NaN   
2200950         NaN         NaN          NaN     NaN                     NaN   
2200951         NaN         NaN          NaN     NaN                     NaN   
2200952         NaN         NaN          NaN     NaN                     NaN   
2200953         NaN         NaN          NaN     NaN                     NaN   

         white blood cell count urine    ph  ph urine  
0                                 NaN  7.40       5.0  
1                                 NaN   NaN       NaN  
2                                 NaN  7.26       NaN  
3                                 NaN   NaN       NaN  
4                                 NaN   NaN       NaN  
...                               ...   ...       ...  
2